daemon/cmd/ipcache.go

```diff
@@ -87,3 +104,126 @@ func containsSubnet(outer, inner net.IPNet) bool {
 
 	return outerBits == innerBits && outerOnes <= innerOnes && outer.Contains(inner.IP)
 }
+
+// restoreIPCache dumps the existing (old) bpf ipcache, adding relevant information
+// back in to the ipcache / identity allocator.
+//
+// For all local (cidr / remote-node) identities, it will add a stub metadata entry
+// that ensures the prefix exists and it gets the same numeric identity before. It
+// reserves those local numeric IDs from allocation to ensure they are available.
+//
+// For ingress IPs, it will add those to the ipcache and configure the local node
+// accordingly.
+//
+// This *must* be called before initMaps(), which will clobber the "old" ipcache.
+func (d *Daemon) restoreIPCache() error {
+	ingressIPs := make([]netip.Prefix, 0, 2)
+	// need to preserve this so we can remove it later.
+	d.restoredCIDRs = map[netip.Prefix]identity.NumericIdentity{}
+
+	// Dump the bpf ipcache, recording any prefixes with local or ingress
+	// numeric identities.
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+
+		if isLocalIdentity(nid) {
```

Comment:
I welcome the refactoring here, but I think it would be wise to split the refactor vs. the functional change into separate commits. At a glance I can see that this conditional check is dropping the other conditionals in the existing logic in the main branch, which makes me nervous that this commit may be breaking other features while doing the refactor + logical change. The simpler the commits are, the lower the likelihood we'll miss something in development & review around how these changes are being applied.

Reply:
I guess arguably if reviewers can pick up on these discrepancies then the patch is not so difficult to follow that it *needs* splitting out, so if it's too much of a pain then let me know and I can take a fresh look at this patch.

Reply:
Also, @christarazi yelled at me for doing too many commits, so I tried to scale back a bit :-)

Reply:
Ah okay, it was just a second-level refactor that I missed during the first pass.

Reply:
I didn't realize I was giving conflicting guidance so sorry about that :) I think it's a good question to grapple with and establish a shared understanding as a community though. Looking at the stats:

```
 daemon/cmd/daemon.go         |  80 +++----------------------------------
 daemon/cmd/daemon_main.go    |  17 +-------
 daemon/cmd/ipcache.go        | 140 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 pkg/ipcache/metadata.go      |  10 -----
 pkg/ipcache/metadata_test.go |  69 +++++++++++++++++---------------
 5 files changed, 184 insertions(+), 132 deletions(-)
```

That's over 300 lines of change in a single patch, which when combining refactors + logical changes can be difficult to keep in your head during review. That's the background to my request. That said, if we don't have agreement amongst reviewers in general then it's unfair for me to push harder on my point here. We can bring this discussion up in general to try to figure out where the line is. Personally I almost exclusively review commit-by-commit and ensure each change is logically separate and does not include sizeable logical and cosmetic changes in the same patch, and I use the patch-by-patch steps as a way to stage my review to reason through the whole thing. Above around 200 lines or so I find that the time to grok a single patch goes from ~single digit minutes to ~double digit minutes as I scan through the diffs in the patch side-by-side.

Anyways, I digress, I will take a fresh look at this patch now I'm more familiar with it.

Reply:
Just chiming in, I don't recall giving commit feedback on this PR and I don't see it as a comment here either :sweat_smile:. The only thing I see regarding commit feedback is https://github.com/cilium/cilium/pull/28673#pullrequestreview-1695783530. If I had given some confusing feedback, I apologize. 

To be clear, I believe having many small and isolated commits as possible makes the reviewer's job easier and thus makes code reviews much more efficient and effective. Keeping refactor commits separate from functional impacting commits is always easier to review and IMO a good practice to evangelize in our community standards. I believe these are standards that we've aligned on in the community AFAICT and I am happy continue contributing in clarifying those and discussing it with the wider community as well.

Back to this PR, I would agree with Joe that splitting out a refactoring commit from the functional change is easier to review. I leave it to you Casey to handle that as you think is best.

Reply:
@christarazi  It's all good :p, it was in a different PR where I was... perhaps a bit excessive with the refactoring commits.

In this case, it wasn't really possible to have a noop refactoring commit. There was logic in a few places within `newDaemon()`, and the need for it to be carefully interleaved only went away when switching to the asynchronous APIs. So I couldn't lift to a single function cleanly -- it would have been a fair bit of throwaway work.